core/src/main/java/com/yahoo/oak/EntryHashSet.java

```diff
@@ -196,55 +207,48 @@ private boolean isKeyAndEntryKeyEqual(KeyBuffer tempKeyBuff, K key, int idx, int
     ** --> ctx.key, ctx.value, ctx.keyHash keep the data of the entry's key, value, and key hash
      */
     private EntryState getEntryState(ThreadContext ctx, int idx, K key, int keyHash) {
+        boolean keyReadResult = false;
+        boolean valueReadResult = false;
 
-        ctx.keyHashAndUpdateCnt = getKeyHashAndUpdateCounter(idx);
+        // optimization, as invalid key reference should be the most frequent case
         if (getKeyReference(idx) == keysMemoryManager.getInvalidReference()) {
             return EntryState.UNKNOWN;
         }
-        // entry was used already, is value valid or deleted?
-        if (getValueReference(idx) != valuesMemoryManager.getInvalidReference()) {
-            // the linearization point of deletion is marking the value off-heap
-            // isValueDeleted checks the reference first (for being deleted) than the off-heap header
-            // isValueDeleted returns true also for invalid reference therefore the check above
-            if (isValueDeleted(ctx.value, idx)) { // value is read to the ctx.value as a side effect
-                // for later progressing with deleted entry read current key slice
-                // (value is read during deleted key check, unless deleted)
-                if (readKey(ctx.key, idx)) {
-                    // key is not deleted: either this deletion is not yet finished,
-                    // or this is a new assignment on top of deleted entry
-                    // Check the key hash:
-                    // if invalid --> this is INSERT_NOT_FINALIZED if valid --> DELETED_NOT_FINALIZED
-                    if (isKeyHashValid(idx)) {
-                        return EntryState.DELETED_NOT_FINALIZED;
+        // try to get constant read of key, value, hash code, its update counter and valid bit
+        do {
+            ctx.keyHashAndUpdateCnt = getKeyHashAndUpdateCounter(idx);
+            // entry is occupied (anyhow), read the key, remember the result
+            keyReadResult = readKey(ctx.key, idx);
+            // read value, if value is invalid or deleted reference the read will fail
+            valueReadResult = readValue(ctx.value, idx);
+        } while (HASH_CODEC.getSecond(ctx.keyHashAndUpdateCnt) != getUpdCntFromHash(idx));
```

Comment:
Why compare just the update counter? You can compare the entire hash?
This would actually do less work since you won't have to extract the counter from the value.

Reply:
agree, fixed